File: src/0206-reverse-linked-list/02.cpp
/*************************************************************************
# File Name: 02.cpp
# Author: Kian Kwok
# Mail: kiankwok6@gmail.com
# Created Time: Mon Mar  2 01:23:38 2020
 ************************************************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode *ans = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        return ans;
    }
};

File: src/0206-reverse-linked-list/02.go
/*************************************************************************
# File Name: 02.go
# Author: Kian Kwok
# Mail: kiankwok6@gmail.com
# Created Time: Wed Nov 25 15:04:26 2020
 ************************************************************************/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    ans := reverseList(head.Next)
    head.Next.Next = head
    head.Next = nil
    return ans
}
